enarx-keep-sev-shim/src/gdt.rs

```diff
@@ -18,8 +18,89 @@ use x86_64::{PrivilegeLevel, VirtAddr};
-pub(crate) static mut GDT: GlobalDescriptorTable = GlobalDescriptorTable::new();
+/// FIXME: will be replaced by a dynamically allocated stack with safe guards
+pub static mut LEVEL_0_STACK: [Page; 5] = [Page::zeroed(); 5];
+
+lazy_static! {
+    /// The global TSS
+    pub static ref TSS: TaskStateSegment = {
+        let mut tss = TaskStateSegment::new();
+
+        #[inline(always)]
+        #[allow(clippy::integer_arithmetic)]
+        fn stack_end(stack_start: VirtAddr, num_pages: usize) -> VirtAddr {
+            stack_start + num_pages * Page::size()
+        }
+
+        unsafe {
+            tss.privilege_stack_table[0] = VirtAddr::from_ptr(&LEVEL_0_STACK);
+        }
+
+        // Assign the stacks for the exceptions and interrupts
+        unsafe {
+            tss.interrupt_stack_table
+                .iter_mut()
+                .zip(&STACKS)
+                .for_each(|(p, v)| *p = stack_end(VirtAddr::from_ptr(v), STACK_NUM_PAGES));
+
+            debug_assert_eq!(STACKS.len(), tss.interrupt_stack_table.len());
+        }
+
+        tss
+    };
+}
+
+/// The Selectors used in the GDT setup
+pub struct Selectors {
+    /// shim code selectore
+    pub code_sel: SegmentSelector,
+    /// shim data selectore
+    pub data_sel: SegmentSelector,
+    /// payload data selector
+    pub user_data_sel: SegmentSelector,
+    /// payload code selector
+    pub user_code_sel: SegmentSelector,
+    /// TSS selector
+    pub tss_sel: SegmentSelector,
+}
+
+lazy_static! {
+    /// The global GDT
+    pub static ref GDT: (GlobalDescriptorTable, Selectors) = {
+        let mut gdt = GlobalDescriptorTable::new();
+
+        // `syscall` loads segments from STAR MSR assuming a data_segment follows `kernel_code_segment`
+        // so the ordering is crucial here. Star::write() will panic otherwise later.
+        let code_sel = gdt.add_entry(Descriptor::kernel_code_segment());
+
+        use DescriptorFlags as Flags;
+        let flags = Flags::USER_SEGMENT | Flags::PRESENT | Flags::WRITABLE;
+
+        let data_sel = gdt.add_entry(Descriptor::UserSegment(flags.bits()));
+
+        // `sysret` loads segments from STAR MSR assuming `user_code_segment` follows `user_data_segment`
+        // so the ordering is crucial here. Star::write() will panic otherwise later.
+        let user_data_sel = gdt.add_entry(Descriptor::user_data_segment());
+        debug_assert_eq!(USER_DATA_SEGMENT, user_data_sel.0 as u64);
+
+        let user_code_sel = gdt.add_entry(Descriptor::user_code_segment());
+        debug_assert_eq!(USER_CODE_SEGMENT, user_code_sel.0 as u64);
+
+        // Important: TSS.deref() != &TSS because of lazy_static
+        let tss_sel = gdt.add_entry(Descriptor::tss_segment(TSS.deref()));
+
+        (
+            gdt,
+            Selectors {
+                code_sel,
+                data_sel,
+                user_data_sel,
+                user_code_sel,
+                tss_sel,
+            },
+        )
```

Comment:
Style nitpick: constructing the `Selectors` struct before the expression would let us avoid the `(` and `)` from being given their own lines. I think it's easier to tell what's going on that way.

```rust
let selectors = Selectors {
    code_sel,
    data_sel,
    user_data_sel,
    user_code_sel,
    tss_sel,
};

(gdt, selectors)

Reply:
hah, cause, it is in a macro, rustfmt doesn't do a format anyway.